internal/lifecycle/lifecycle.go

```diff
@@ -74,18 +101,105 @@ func (l *Lifecycle) Start(ctx context.Context) error {
 // hooks run in reverse order.
 func (l *Lifecycle) Stop(ctx context.Context) error {
 	var errs []error
+	l.mu.Lock()
+	l.records = make(HookRecords, 0, l.numStarted)
+	l.mu.Unlock()
+
 	// Run backward from last successful OnStart.
 	for ; l.numStarted > 0; l.numStarted-- {
 		hook := l.hooks[l.numStarted-1]
 		if hook.OnStop == nil {
 			continue
 		}
-		l.logger.LogEvent(&fxevent.LifecycleHookStop{CallerName: hook.caller})
+
+		l.logger.LogEvent(&fxevent.LifecycleHookStop{CallerName: hook.callerFrame.Function})
+
+		l.mu.Lock()
+		l.runningHook = hook
+		l.mu.Unlock()
+
+		begin := time.Now()
 		if err := hook.OnStop(ctx); err != nil {
 			// For best-effort cleanup, keep going after errors.
 			errs = append(errs, err)
 		}
+		l.mu.Lock()
+		l.records = append(l.records, HookRecord{
+			CallerFrame: hook.callerFrame,
+			Func:        hook.OnStop,
+			Runtime:     time.Since(begin),
+		})
+		l.mu.Unlock()
 	}
 
 	return multierr.Combine(errs...)
 }
+
+// HookRecords returns the info of hooks that successfully ran till the end,
+// including their caller and runtime. Used to report timeout errors on Start/Stop.
+func (l *Lifecycle) HookRecords() HookRecords {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+	r := make(HookRecords, len(l.records))
+	copy(r, l.records)
+	return r
+}
+
+// RunningHookCaller returns the name of the hook that was running when a Start/Stop
+// hook timed out.
+func (l *Lifecycle) RunningHookCaller() string {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+	return l.runningHook.callerFrame.Function
+}
+
+// HookRecord keeps track of each Hook's execution time, the caller that appended the Hook, and function that ran as the Hook.
+type HookRecord struct {
+	CallerFrame fxreflect.Frame             // stack frame of the caller
+	Func        func(context.Context) error // function that ran as sanitized name
+	Runtime     time.Duration               // how long the hook ran
+}
+
+// HookRecords is a Stringer wrapper of HookRecord slice.
+type HookRecords []HookRecord
+
+func (rs HookRecords) Len() int {
+	return len(rs)
+}
+
+func (rs HookRecords) Less(i, j int) bool {
+	// Sort by runtime, greater ones at top.
+	return rs[i].Runtime > rs[j].Runtime
+}
+
+func (rs HookRecords) Swap(i, j int) {
+	rs[i], rs[j] = rs[j], rs[i]
+}
+
+// Used for logging startup errors.
+func (rs HookRecords) String() string {
+	var b strings.Builder
+	for _, r := range rs {
+		fmt.Fprintf(&b, "%s took %v from %s",
+			fxreflect.FuncName(r.Func), r.Runtime, r.CallerFrame)
+	}
+	return b.String()
+}
+
+// Format implements fmt.Formatter to handle "%+v".
+func (rs HookRecords) Format(w fmt.State, c rune) {
+	if !w.Flag('+') {
+		// Without %+v, fall back to String().
+		io.WriteString(w, rs.String())
+		return
+	}
+
+	sort.Slice(rs, func(i, j int) bool { return rs[i].Runtime > rs[j].Runtime })
+	for _, r := range rs {
+		fmt.Fprintf(w, "\n%s took %v from:\n\t%+v",
+			fxreflect.FuncName(r.Func),
+			r.Runtime,
+			r.CallerFrame)
+	}
+	fmt.Printf("\n")
```

Comment:
Did you mean Fprintf.
```suggestion
	fmt.Fprintf(w, "\n")
```